From bea33a586768af4e47aeed99519816733a8940a5 Mon Sep 17 00:00:00 2001
From: alvindimas05 <alvin.dimas.praditya@gmail.com>
Date: Fri, 23 Oct 2026 10:24:09 +0700
Subject: [PATCH 20/35] [memory_manager] merge translation caches into single
 thread-local struct

(cherry picked from commit e16b40afe25d2829f4f37ce18ef3ec685f3f937e)
---
 src/video_core/memory_manager.cpp | 11 +++++++----
 1 file changed, 7 insertions(+), 4 deletions(-)

diff --git a/src/video_core/memory_manager.cpp b/src/video_core/memory_manager.cpp
index e9cf015cf..6f3b064a0 100644
--- a/src/video_core/memory_manager.cpp
+++ b/src/video_core/memory_manager.cpp
@@ -223,8 +223,11 @@ std::optional<DAddr> MemoryManager::GpuToCpuAddress(GPUVAddr gpu_addr) const {
         bool mapped{};
     };
     static constexpr size_t kCacheEntries = 2048;
-    static thread_local std::array<CacheEntry, kCacheEntries> big_cache{};
-    static thread_local std::array<CacheEntry, kCacheEntries> short_cache{};
+    struct TranslateCache {
+        std::array<CacheEntry, kCacheEntries> big{};
+        std::array<CacheEntry, kCacheEntries> short_table{};
+    };
+    static thread_local TranslateCache cache{};
 
     const u32 gen_start{translation_generation.load(std::memory_order_acquire)};
     if ((gen_start & 1) != 0) [[unlikely]] {
@@ -241,7 +244,7 @@ std::optional<DAddr> MemoryManager::GpuToCpuAddress(GPUVAddr gpu_addr) const {
 
     const size_t big_index{PageEntryIndex(gpu_addr, true)};
     const size_t big_slot{(big_index ^ (big_index >> 11)) & (kCacheEntries - 1)};
-    CacheEntry& big_entry{big_cache[big_slot]};
+    CacheEntry& big_entry{cache.big[big_slot]};
     if (big_entry.instance == unique_identifier && big_entry.generation == gen_start &&
         big_entry.index == big_index) {
         if (big_entry.mapped) {
@@ -262,7 +265,7 @@ std::optional<DAddr> MemoryManager::GpuToCpuAddress(GPUVAddr gpu_addr) const {
 
     const size_t short_index{PageEntryIndex(gpu_addr, false)};
     const size_t short_slot{(short_index ^ (short_index >> 11)) & (kCacheEntries - 1)};
-    CacheEntry& short_entry{short_cache[short_slot]};
+    CacheEntry& short_entry{cache.short_table[short_slot]};
     if (short_entry.instance == unique_identifier && short_entry.generation == gen_start &&
         short_entry.index == short_index) {
         if (!short_entry.mapped) {
-- 
2.43.0

